src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/zip/UnZipStepExecution.java

```diff
@@ -144,17 +147,12 @@ public UnZipFileCallable(TaskListener listener, FilePath destination, String glo
                         } else {
                             logger.print("Reading: ");
                             logger.println(entry.getName());
-                            try (BufferedReader reader = new BufferedReader(new InputStreamReader(zip.getInputStream(entry), Charset.defaultCharset()))) {
-                                String line = reader.readLine();
-                                StringBuilder str = new StringBuilder();
-                                while (line != null) {
-                                    if (str.length() >= 0) {
-                                        str.append('\n');
-                                    }
-                                    str.append(line);
-                                    line = reader.readLine();
-                                }
-                                strMap.put(entry.getName(), str.toString().trim());
+                            // we need to copy byte by byte everything to be sure that no carriage return characters are skipped
+                            // readLine skips the carriage return and for example files ending with only one carriage return are trimmed
+
+                            try (InputStream is = zip.getInputStream(entry); ByteArrayOutputStream output = new ByteArrayOutputStream()) {
+                                IOUtils.copyLarge(is, output);
+                                strMap.put(entry.getName(), new String(output.toByteArray(), Charset.defaultCharset()));
```

Comment:
You think it might be better to fix to UTF-8? I have never worked with ZipFiles so I used the same approach used previously. What I have read from JavaDoc is that `The UTF-8 charset is used to decode the entry names and comments.` so it seems that the content itself would be in the original encode right?